clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -3532,6 +3535,117 @@ check_has_instructions(void)
+            return false;
+    }
+
+    // Incorrect: TRACE_MARKER_TYPE_SYSCALL not allowed.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FILETYPE,
+                       OFFLINE_FILE_TYPE_SYSCALL_NUMBERS |
+                           OFFLINE_FILE_TYPE_ARCH_REGDEPS),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_SYSCALL, 102),
+            gen_instr(TID_A),
+            gen_exit(TID_A),
+        };
+        if (!run_checker(memrefs, true,
+                         { "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                           "TRACE_MARKER_TYPE_SYSCALL markers",
+                           /*tid=*/TID_A,
+                           /*ref_ordinal=*/4, /*last_timestamp=*/0,
+                           /*instrs_since_last_timestamp=*/0 },
+                         "Failed to catch non-allowed TRACE_MARKER_TYPE_SYSCALL marker"))
+            return false;
+    }
+
+#ifdef LINUX
+    // Incorrect: non SYS_futex TRACE_MARKER_TYPE_FUNC_ID not allowed.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ARCH_REGDEPS),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FUNC_ID, 102),
+            gen_instr(TID_A),
+            gen_exit(TID_A),
+        };
+        if (!run_checker(memrefs, true,
+                         { "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                           "TRACE_MARKER_TYPE_FUNC_ID markers related to functions that "
+                           "are not SYS_futex",
+                           /*tid=*/TID_A,
+                           /*ref_ordinal=*/4, /*last_timestamp=*/0,
+                           /*instrs_since_last_timestamp=*/0 },
+                         "Failed to catch non-allowed TRACE_MARKER_TYPE_FUNC_ID marker"))
+            return false;
+    }
+
+    // Correct: SYS_futex TRACE_MARKER_TYPE_FUNC_ID allowed.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ARCH_REGDEPS),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+            gen_marker(TID_A, TRACE_MARKER_TYPE_FUNC_ID,
+                       static_cast<uintptr_t>(func_trace_t::TRACE_FUNC_ID_SYSCALL_BASE) +
+                           SYS_futex),
+            gen_instr(TID_A),
+            gen_exit(TID_A),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+#else
+    // Incorrect: TRACE_MARKER_TYPE_FUNC_ID not allowed when function ID cannot be
+    // determined.
```

Comment:
?  Why couldn't we determine the ID?

Reply:
My rationale here was: let's say someone has a regdeps trace captured on linux with SYS_futex function markers, if they run the invariant_checker on that trace on, let's say, a windows platform, then that build of invariant_checker won't have SYS_futex, so it cannot be determined if that TRACE_MARKER_TYPE_FUNC_ID is allowed or not (hence the different error message).
However, now that I read it again, it should have been something like: "We cannot determine if the function ID of this TRACE_MARKER_TYPE_FUNC_ID is allowed in an OFFLINE_FILE_TYPE_ARCH_REGDEPS trace".
Does it make sense?
I'm also fine with a single error message as suggested, I just thought it would be more descriptive.

Reply:
Oh, wasn't thinking of cross-OS use.  In fact: we don't store the OS as part of the trace type so even the ifdef-LINUX could be looking at a Windows trace whose syscall # happens to match SYS_futex.

Maybe just remove the ifdefs and allow SYS_futex for the arch regardless of the OS (presumably can get syscall # from our headers even if non-Linux build)?  Seems low risk.

Reply:
Oh, right, the opposite (windows trace on a linux build) can also happen, and yes if the windows syscall matches SYS_futex, that invariant check would pass.

Unfortunately we can't completely drop the #ifdef, or a windows build of invariant_checker would fail.
SYS_futex comes from `core/unix/include/syscall_target.h`, which would break the build if included when LINUX is not defined (https://github.com/DynamoRIO/dynamorio/blob/577db95d9d0741844b59fe24c50be61aefe276d8/core/unix/include/syscall_target.h#L6).
We could redefine SYS_futex to some value if LINUX is not defined, but it looks a bit ugly in my opinion.

Let's simplify and drop the two different messages, the original one is not wrong after all.
In non-LINUX builds of DR the check will always fail, which is the same behavior as having #syscall != SYS_futex.

Reply:
Hmm, will it be annoying that someone building analysis tools on Windows to process public traces has this invariant fail every time when really there's nothing wrong with the trace?  Should we go the other direction and let it pass (maybe with a warning)?  After all, the main point of these new checks is for us internally to make sure a to-be-released trace has the contents we expect.  Externally, there's no problem with a trace that has some extra syscall or where you can't tell whether it has SYS_futex or something else: the trace is still logically consistent.

Reply:
Agreed. Done (also adjusted the unit test to reflect that).